### DISCUSSION TITLE
How to use Util.telescope from your own lua files

### DISCUSSION THREAD

**Discussion by kevclark** [LazyVim/LazyVim#338] [Q&A] [answered]
I'd like to override `ff` and `fF` maps and pass in the explicit `find_files` and `git_files` params, as I don't find it that useful to let lazyvim auto switch to the git view if you happen to be in a git repo.  An example being in a large repo, buried several directories deep, and you just want to do a fuzzy find from that level.  Which may include auto generated files from a build unknown to git.

Anyway, I'd still like to make use of the LazyVim Util helper, as it has the find root functionality.  However, when trying to require the Util with a new key map, it does nothing.  No errors, just does not bring up the finder.  

Is it not possible to require Util outside of LazyVim?

```
      -- override default LazyVim find files maps
      {
        "<leader>ff",
        function()
          local Util = require("lazyvim.util")
          Util.telescope("find_files")
        end,
        desc = "Find Files (root dir)",
      }
```

**Comment by dpetka2001** [comment]
Of course you can stil use it like you did in your own config. It's just that the `Util.telescope` has inherent support for `.git` files, so if you're not interested in that, just use traditional `<leader>ff <cmd>lua require('telescope.builtin').find_files()<cr>` or `<leader>ff <cmd>Telescope find_files<cr>` and use `Util.get_root()` to find the `root_dir` of whatever you want. You can't expect to use `Util.telescope` and not use support for `.git` files because it's inherent in the function definition.

**Reply by kevclark** [reply]
As mentioned above, calls to Util.* don't appear to be working for me.  So yes, I realise I could require telescope directly, but there's little point if calling it is already wrapped by Util.  My question relates to why calls to Util do not appear to be working for me.

**Reply by dpetka2001** [reply]
What exactly doesn't work for you? I use it in my keymaps to change the size of lazygit window and also in my autocmd file to create a user command and it works just fine.

**Reply by dpetka2001** [reply]
If you use `Util.telescope` in your keymappings you can't expect it to seperate actions between `find_files` and `git_files` because the function definition does not do that. It takes into account both `git_files` and `find_files`. If you want 2 seperate keymappings then you should do it as I said.

**Reply by kevclark** [reply]
I don't expect that, I would add two separate mappings for both `find_files` and `git_files`, sorry I thought that would have been obvious to where I was going.  If you look at the definition of `Util.telescope` you will see that the auto switching only occurs if the param is `files`, otherwise it will pass whatever you supply straight through to `require("telescope.builtin")[builtin](opts)`.  My point is, when I call this with the code in the opening question, it does nothing.  Even if I pass the default param that lazyvim uses `files`.

**Comment by dpetka2001** [comment, marked as answer]
```lua
local Util = require("lazyvim.util")
return {
  {
    "telescope.nvim",
    keys = {
      { "<leader>ff", Util.telescope("find_files"), desc = "Find files (root dir) },
    },
  },
}
```
Could you check this out? Seems to work for me.

Edit: Updated `Util` to be local.

**Comment by kevclark** [comment]
Interesting, yes that does work, thank you!  Thought I had already tried this approach, which is the same way that `lazyvim/plugins/editor.lua` does it.  Not really sure why not requiring the function directly within the keymap should not work though.
```
      {
        "<leader>ff",
        function()
          -- local Util = require("lazyvim.util")
          -- Util.telescope("find_files")
          require("lazyvim.util").telescope("find_files")
        end,
        desc = "Find Files (root dir)",
      }
```

**Reply by dpetka2001** [reply]
I'm not sure entirely myself, but my guess would be that when you require the `Utils` in the function key, it's loaded too late and doesn't take effect when you call it maybe?

**Reply by folke** [reply]
It's because that Util.telescope returns a function, so in your example, you would need to call it

`require("lazyvim.util").telescope("find_files")()`.

See the added `()` at the end

**Reply by dpetka2001** [reply]
Oh I see thnx for your explanation and welcome back :smiley:
